### PR TITLE
Orca SparkXShard support broadcast variables in transform_shard

### PR DESCRIPTION
**Pull request by cyita**

### CODE REVIEW

File: pyzoo/zoo/orca/data/shard.py [hunk 1]
```diff
@@ -130,8 +130,14 @@ def __init__(self, rdd):
         if self.eager:
             self.compute()
 
-    def transform_shard(self, func, *args):
-        transformed_shard = SparkXShards(self.rdd.map(lambda data: func(data, *args)))
+    def transform_shard(self, func, broadcast_args=False, sc=None, *args):
+        if broadcast_args:
+            assert sc is not None, "Spark context should be provided"
+            broadcast_args_value = tuple(map(lambda arg: sc.broadcast(arg).value, args))
```
Comment (qiuxin2012) [inline pyzoo/zoo/orca/data/shard.py:136]: you should call `.value`  in `rdd.map`

File: pyzoo/zoo/orca/data/shard.py [hunk 2]
```diff
@@ -133,9 +133,11 @@ def __init__(self, rdd):
     def transform_shard(self, func, broadcast_args=False, sc=None, *args):
         if broadcast_args:
             assert sc is not None, "Spark context should be provided"
-            broadcast_args_value = tuple(map(lambda arg: sc.broadcast(arg).value, args))
+            broadcast_args = tuple(map(lambda arg: sc.broadcast(arg), args))
```
Comment (qiuxin2012) [inline pyzoo/zoo/orca/data/shard.py:136]: Does this work?
```
            broadcast_args = sc.broadcast(args)
            transformed_shard = SparkXShards(self.rdd.map(lambda data: func(data,
                                                                            *broadcast_args.value)))
```

Comment (cyita) [reply pyzoo/zoo/orca/data/shard.py]: It works. Fixed.

File: pyzoo/zoo/orca/data/shard.py (hunk 1 above)
Comment (jenniew) [reply pyzoo/zoo/orca/data/shard.py]: do we need to pass sc or just use self.rdd.ctx?

Comment (qiuxin2012) [reply pyzoo/zoo/orca/data/shard.py]: Yes, I think we should use rdd's sc.

**Comment by jason-dai** [comment]
I would prefer something like:

```
def func(df, item_set)
   item_set = item_set.value
   ....

item_set= ...
item_set= orca.data.SharedValue(item_set)
full_data.transform_shard(func, item_set)
```

**Comment by jason-dai** [comment]
Alternatively, we can just broadcast the variables by default; not sure about the performance implications for smaller objects.

File: pyzoo/zoo/orca/data/shard.py [hunk 3]
```diff
@@ -280,3 +281,20 @@ def load_pickle(cls, path, minPartitions=None):
 
     def __del__(self):
         self.uncache()
+
+
+class SharedValue(object):
+    def __init__(self, data, sc):
+        if isinstance(sc, SparkContext):
+            self.broadcast_data = sc.broadcast(data)
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/shard.py:289]: just use `sc = init_nncontext()`, similar to https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/orca/data/shard.py#L313

**Comment by cyita** [comment]
Jenkins passed: http://10.239.47.210:18888/view/ZOO-PR-PIPELINE/job/ZOO-PR-Validation/3586/

**Pull request merged**